Access control: initial part prepared for merging

alexanderzobnin opened the pull request:
Reorganizing packages according to [Ben Johnson](https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1) Standard Package Layout.
Rename RBAC -> access control

alexanderzobnin committed:
RBAC: database structure

alexanderzobnin committed:
Convert to the service

alexanderzobnin committed:
More db queries

alexanderzobnin committed:
basic middleware

alexanderzobnin committed:
Add/remove database actions

alexanderzobnin committed:
Initial DB tests

alexanderzobnin committed:
Evaluation tests WIP

alexanderzobnin committed:
Define access control service interface

alexanderzobnin committed:
Refactor database structure

alexanderzobnin committed:
Add builtin_role_policy table

alexanderzobnin committed:
Fix builtin_role_policy role index

alexanderzobnin committed:
Refactor permission table

alexanderzobnin committed:
Simplify permissions query

sakjur committed:
Work on authZ middleware

- Moved to its own package
- Add fake authorizer
- Rename some things
- Don't apply to views just yet (only APIs)

alexanderzobnin committed:
Tests for getting user policies from teams

alexanderzobnin committed:
Get user permissions from teams

alexanderzobnin committed:
Fetch all user policies (from teams as well)

alexanderzobnin committed:
Fix policies search

alexanderzobnin committed:
Update policies

sakjur committed:
Test globs for scope matching and templating for parameters

sakjur committed:
Merge remote-tracking branch 'origin/master' into access-control-stage1

# Conflicts:
#	pkg/api/http_server.go

alexanderzobnin committed:
Fix policies and permissions queries

sakjur committed:
Middleware: Move to package

sakjur committed:
Evaluation: Start work on database backed evaluator

sakjur committed:
access control/database: prefer return values instead of mutation

sakjur committed:
Start work on database seeder

alexanderzobnin committed:
database: prefer return values instead of mutation

alexanderzobnin committed:
database: benchmarks for policies

alexanderzobnin committed:
script for graphing benchmark results

alexanderzobnin committed:
Fix querying policies

alexanderzobnin committed:
Fix benchmark viz script

alexanderzobnin committed:
Benchmark: put commit hash to the graph title

alexanderzobnin committed:
Benchmark: build 2 charts

alexanderzobnin committed:
Benchmark: separate users and policies

sakjur committed:
Make database operations cancellable

sakjur committed:
Add seeder behind feature toggle

sakjur committed:
Use same method to set up environment for benchmarks + tests

alexanderzobnin committed:
Initial access control API

alexanderzobnin committed:
Use UID for policies

alexanderzobnin committed:
Move API to enterprise

alexanderzobnin committed:
Merge remote-tracking branch 'origin/master' into access-control-stage1

alexanderzobnin committed:
Delete policy both by UID and ID

alexanderzobnin committed:
Make errors available for external packages

sakjur committed:
Access Control: Seeding tests

alexanderzobnin committed:
Create policy with permissions

alexanderzobnin committed:
Update policy with permissions

alexanderzobnin committed:
API: update policy with permissions

sakjur committed:
Seeder: Add role mappings

sakjur committed:
Seeder: refactoring

sakjur committed:
Seeder: improve testing

sakjur committed:
Use permissions database query

sakjur committed:
Remove org permissions + scopes

sakjur committed:
Start working on built in role support

alexanderzobnin committed:
Create policy with UID

alexanderzobnin committed:
tests: use consistent permission naming

alexanderzobnin committed:
database: make org_id_uid index unique

sakjur committed:
Merge branch 'master' into access-control-stage1

sakjur committed:
Merge branch 'master' into access-control-stage1

sakjur committed:
Add support for builtin roles in permissions and policy query

sakjur committed:
Inherit built-in roles

sakjur committed:
seeder: fix error being thrown erroneously

sakjur committed:
Merge branch 'master' into access-control-stage1

sakjur committed:
Fix merge

sakjur committed:
reorganize imports

sakjur committed:
add support for policy versions

sakjur committed:
Use policy versions for seeding

sakjur committed:
Update seeder test

sakjur committed:
Disable entire feature and database migrations when no feature toggle

sakjur committed:
Move evaluator to separate file

alexanderzobnin committed:
Refactor: reorganize access control package

sakjur commented on pkg/api/api.go:
```diff
@@ -12,12 +12,14 @@ import (
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/middleware"
 	"github.com/grafana/grafana/pkg/models"
+	acmiddleware "github.com/grafana/grafana/pkg/services/accesscontrol/middleware"
```
We should remove all changes to `api.go` before merging, that's not ready to be merged.

alexanderzobnin replied:
Sure, forget about that doing the refactor.

sakjur commented on pkg/api/http_server.go:
```diff
@@ -13,12 +13,9 @@ import (
-	"github.com/grafana/grafana/pkg/services/shorturls"
-	"github.com/grafana/grafana/pkg/services/sqlstore"
-	"github.com/grafana/grafana/pkg/tsdb"
+	"github.com/prometheus/client_golang/prometheus"
```
We can remove all changes to `http_server.go` as well.

sakjur commented on pkg/services/accesscontrol/manager/manager.go:
```diff
@@ -0,0 +1,53 @@
+}
+
+func (ac *AccessControlService) IsDisabled() bool {
+	_, exists := ac.Cfg.FeatureToggles["new_authz"]
```
```suggestion
	_, exists := ac.Cfg.FeatureToggles["accesscontrol"]
```

I never quite liked my previous name for this 😇

sakjur reviewed (commented):
Just want to remove some things I've left in the other branch that would cause problems for people if it got to master, but after that I think it'd be nice to get this in and start working on smaller sets of changes

alexanderzobnin committed:
Refactor: rename policy -> role

alexanderzobnin committed:
Refactor: cleanup

alexanderzobnin committed:
Remove using access control

alexanderzobnin committed:
Rename feature toggle to accesscontrol

alexanderzobnin commented on pkg/services/accesscontrol/accesscontrol.go:
```diff
@@ -0,0 +1,44 @@
+	RemoveTeamRole(cmd *RemoveTeamRoleCommand) error
+	AddUserRole(cmd *AddUserRoleCommand) error
+	RemoveUserRole(cmd *RemoveUserRoleCommand) error
+	AddBuiltinRoleRole(ctx context.Context, orgID, roleID int64, roleName string) error
```
Probably, this better
```suggestion
	AddBuiltinRole(ctx context.Context, orgID, roleID int64, roleName string) error
```

sakjur replied:
Yes :joy:

alexanderzobnin committed:
Fix linter errors

alexanderzobnin committed:
FIx AddBuiltinRole() name

alexanderzobnin committed:
Fix shellcheck

alexanderzobnin committed:
Fix shellcheck

aknuds1 commented on pkg/services/accesscontrol/manager/manager_test.go:
```diff
@@ -0,0 +1,116 @@
+
+			res, err := ac.GetUserRoles(context.Background(), userRolesQuery)
+			require.NoError(t, err)
+			assert.Equal(t, len(tc.roles), len(res))
```
```suggestion
			assert.Len(t, res, len(tc.roles))
```

aknuds1 commented on pkg/services/accesscontrol/manager/manager_test.go:
```diff
@@ -0,0 +1,116 @@
+
+			permissions, err := ac.GetUserPermissions(context.Background(), userPermissionsQuery)
+			require.NoError(t, err)
+			assert.Equal(t, len(tc.roles[0].Permissions), len(permissions))
```
```suggestion
			assert.Len(t, permissions, len(tc.roles[0].Permissions))
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+
+// Role is the model for Role in RBAC.
+type Role struct {
+	Id          int64  `json:"id"`
+	OrgId       int64  `json:"orgId"`
```
In Go, initialisms should be written in consistent case:
```suggestion
	ID          int64  `json:"id"`
	OrgID       int64  `json:"orgId"`
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type RoleDTO struct {
+	Id          int64        `json:"id"`
+	OrgId       int64        `json:"orgId"`
```
```suggestion
	ID          int64        `json:"id"`
	OrgID       int64        `json:"orgId"`
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+
+// Permission is the model for Permission in RBAC.
+type Permission struct {
+	Id         int64  `json:"id"`
+	RoleId     int64  `json:"-"`
```
```suggestion
	ID         int64  `json:"id"`
	RoleID     int64  `json:"-"`
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type TeamRole struct {
+	Id     int64
+	OrgId  int64
+	RoleId int64
+	TeamId int64
```
```suggestion
	ID     int64
	OrgID  int64
	RoleID int64
	TeamId int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type UserRole struct {
+	Id     int64
+	OrgId  int64
+	RoleId int64
+	UserId int64
```
```suggestion
	ID     int64
	OrgID  int64
	RoleID int64
	UserId int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type GetTeamRolesQuery struct {
+	OrgId  int64 `json:"-"`
+	TeamId int64
```
```suggestion
	OrgID  int64 `json:"-"`
	TeamID int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type GetUserRolesQuery struct {
+	OrgId  int64 `json:"-"`
+	UserId int64
```
```suggestion
	OrgID  int64 `json:"-"`
	UserID int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type GetUserPermissionsQuery struct {
+	OrgId  int64 `json:"-"`
+	UserId int64
```
```suggestion
	OrgID  int64 `json:"-"`
	UserID int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type CreatePermissionCommand struct {
+	RoleId     int64
```
```suggestion
	RoleID     int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type UpdatePermissionCommand struct {
+	Id         int64
```
```suggestion
	ID         int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type DeletePermissionCommand struct {
+	Id int64
```
```suggestion
	ID int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type CreateRoleCommand struct {
+	OrgId       int64  `json:"-"`
```
```suggestion
	OrgID       int64  `json:"-"`
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type CreateRoleWithPermissionsCommand struct {
+	OrgId       int64        `json:"orgId"`
```
```suggestion
	OrgID       int64        `json:"orgId"`
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type UpdateRoleCommand struct {
+	Id          int64        `json:"id"`
+	OrgId       int64        `json:"orgId"`
```
```suggestion
	ID          int64        `json:"id"`
	OrgID       int64        `json:"orgId"`
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type DeleteRoleCommand struct {
+	Id    int64
```
```suggestion
	ID    int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+type DeleteRoleCommand struct {
+	Id    int64
+	UID   string `json:"uid"`
+	OrgId int64
```
```suggestion
	OrgID int64
```

aknuds1 commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+type AddTeamRoleCommand struct {
+	OrgId  int64
+	RoleId int64
+	TeamId int64
+}
+
+type RemoveTeamRoleCommand struct {
+	OrgId  int64
+	RoleId int64
+	TeamId int64
+}
+
+type AddUserRoleCommand struct {
+	OrgId  int64
+	RoleId int64
+	UserId int64
+}
+
+type RemoveUserRoleCommand struct {
+	OrgId  int64
+	RoleId int64
+	UserId int64
```
```suggestion
	OrgID  int64
	RoleID int64
	TeamID int64
}

type RemoveTeamRoleCommand struct {
	OrgID  int64
	RoleID int64
	TeamID int64
}

type AddUserRoleCommand struct {
	OrgID  int64
	RoleID int64
	UserID int64
}

type RemoveUserRoleCommand struct {
	OrgID  int64
	RoleID int64
	UserID int64
```

aknuds1 commented on pkg/services/accesscontrol/middleware/middleware.go:
```diff
@@ -0,0 +1,45 @@
+
+				tmpl, err := template.New("scope").Parse(scope)
+				if err != nil {
+					c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
```
You should return here?
```suggestion
					c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
					return
```

aknuds1 commented on pkg/services/accesscontrol/middleware/middleware.go:
```diff
@@ -0,0 +1,45 @@
+				}
+				err = tmpl.Execute(&buf, c.AllParams())
+				if err != nil {
+					c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
```
Looks as if you should return:

```suggestion
					c.JsonApiErr(http.StatusInternalServerError, "Internal server error", err)
					return
```

aknuds1 reviewed (dismissed):
Generally looks solid, but I think I spotted a couple of error checks that should return? Also various nits.

sakjur commented on pkg/services/sqlstore/rbac.go:
```diff
@@ -0,0 +1 @@
+package sqlstore
```
This file can be deleted, right?

sakjur committed:
Apply suggestions from code review

(except for rename Id->ID as those will require changes elsewhere and maybe the addition of tags)

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

sakjur committed:
Apply review comments: Id->ID

sakjur committed:
Delete empty file

aknuds1 commented on pkg/services/accesscontrol/manager/manager_test.go:
```diff
@@ -0,0 +1,117 @@
+	return ac
+}
+
+func setupTestEnv(t testing.TB) *AccessControlService {
```
I think these test utility functions should start with `t.Helper()`.
```suggestion
func setupTestEnv(t testing.TB) *AccessControlService {
	t.Helper()

```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+	return &store
+}
+
+func OverrideDatabaseInRegistry(cfg *setting.Cfg) accessControlStoreTestImpl {
```
```suggestion
func overrideDatabaseInRegistry(cfg *setting.Cfg) accessControlStoreTestImpl {
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+func setupTestEnv(t testing.TB) *accessControlStoreTestImpl {
+	cfg := setting.NewCfg()
+
+	store := OverrideDatabaseInRegistry(cfg)
```
```suggestion
	store := overrideDatabaseInRegistry(cfg)
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+
+func TestCreatingRole(t *testing.T) {
+	MockTimeNow()
+	defer ResetTimeNow()
```
```suggestion
	t.Cleanup(ResetTimeNow)
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+			if tc.role.Permissions == nil {
+				assert.Empty(t, role.Permissions)
+			} else {
+				assert.Equal(t, len(tc.role.Permissions), len(role.Permissions))
```
```suggestion
				assert.Len(t, role.Permissions, len(tc.role.Permissions))
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+
+func TestUpdatingRole(t *testing.T) {
+	MockTimeNow()
+	defer ResetTimeNow()
```
```suggestion
	t.Cleanup(ResetTimeNow)
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+			_, err := store.UpdateRole(context.Background(), updateRoleCmd)
+			require.NoError(t, err)
+
+			updatedRole, err := store.GetRoleByUID(context.Background(), 1, role.UID)
+
+			require.NoError(t, err)
```
```suggestion
			updatedRole, err := store.GetRoleByUID(context.Background(), 1, role.UID)
			require.NoError(t, err)
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+
+			createRoleRes := actesting.CreateRole(t, store, tc.role)
+
+			res, err := store.GetRoleByUID(context.Background(), 1, createRoleRes.UID)
+			role := res
+
+			require.NoError(t, err)
```
```suggestion
			role, err := store.GetRoleByUID(context.Background(), 1, createRoleRes.UID)
			require.NoError(t, err)
```

aknuds1 commented on pkg/services/accesscontrol/database/database_test.go:
```diff
@@ -0,0 +1,452 @@
+
+			res, err := store.GetUserPermissions(context.Background(), userPermissionsQuery)
+			require.NoError(t, err)
+			assert.Equal(t, len(expectedPermissions), len(res))
```
```suggestion
			assert.Len(t, res, len(expectedPermissions))
```

aknuds1 commented on pkg/services/accesscontrol/manager/manager_test.go:
```diff
@@ -0,0 +1,117 @@
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+func overrideAccessControlInRegistry(cfg *setting.Cfg) AccessControlService {
```
I think these test utility functions should start with t.Helper().
```suggestion
func overrideAccessControlInRegistry(t *testing.TB, cfg *setting.Cfg) AccessControlService {
	t.Helper()

```

aknuds1 commented on pkg/services/accesscontrol/manager/manager_test.go:
```diff
@@ -0,0 +1,117 @@
+
+func TestEvaluatingPermissions(t *testing.T) {
+	database.MockTimeNow()
+	defer database.ResetTimeNow()
```
```suggestion
	t.Cleanup(database.ResetTimeNow)
```

aknuds1 commented on pkg/services/accesscontrol/seeder/seeder.go:
```diff
@@ -0,0 +1,219 @@
+
+	existingPermissions, err := s.Store.GetRolePermissions(ctx, old.ID)
+	if err != nil {
+		s.log.Info("failed to get current permissions for role", "name", role.Name, "err", err)
```
Is it OK to continue here?

alexanderzobnin replied:
@sakjur ?

sakjur replied:
🤔 I think both of these should be returned. I'll change that.

aknuds1 commented on pkg/services/accesscontrol/seeder/seeder.go:
```diff
@@ -0,0 +1,219 @@
+
+	err = s.idempotentUpdatePermissions(ctx, old.ID, role.Permissions, existingPermissions)
+	if err != nil {
+		s.log.Error("failed to update role permissions", "name", role.Name, "err", err)
```
The error shouldn't be returned?

aknuds1 commented on pkg/services/accesscontrol/seeder/seeder_test.go:
```diff
@@ -0,0 +1,165 @@
+	database.AccessControlStore
+}
+
+func init() {
```
Can we avoid using `init` for tests? It's really dirty.

aknuds1 commented on pkg/services/accesscontrol/seeder/seeder_test.go:
```diff
@@ -0,0 +1,165 @@
+	database.AddAccessControlMigrations(mg)
+}
+
+func setupTestEnv(t testing.TB) *accessControlStoreTestImpl {
+	cfg := setting.NewCfg()
```
```suggestion
func setupTestEnv(t testing.TB) *accessControlStoreTestImpl {
	t.Helper()

	cfg := setting.NewCfg()
```

aknuds1 commented on pkg/services/accesscontrol/seeder/seeder_test.go:
```diff
@@ -0,0 +1,165 @@
+	return &store
+}
+
+func overrideDatabaseInRegistry(cfg *setting.Cfg) accessControlStoreTestImpl {
+	store := accessControlStoreTestImpl{
```
```suggestion
func overrideDatabaseInRegistry(t *testing.TB, cfg *setting.Cfg) accessControlStoreTestImpl {
	t.Helper()
	
	store := accessControlStoreTestImpl{
```

aknuds1 reviewed (commented):
Generally looks good, but please see comments.

alexanderzobnin committed:
Chore: apply review suggestions

alexanderzobnin committed:
Chore: simplify seeder tests

alexanderzobnin committed:
Chore: simplify tests

sakjur committed:
Seeder: Fail on database error

sakjur committed:
Merge branch 'master' into access-control-merge-to-master

# Conflicts:
#	pkg/api/http_server.go

sakjur committed:
Fix merge

sakjur committed:
Fix tests with new create user methods

sakjur reviewed (approved):
Since I've written so much of this, I'll ask for another round of feedback before merging, but I'd rather have this merged and work on any feedback that isn't blocking in separate PRs to not have to maintain a long living WIP branch and risk growing this even more 😇

wbrowne commented on pkg/services/accesscontrol/manager/evaluator.go:
```diff
@@ -0,0 +1,72 @@
+			}
+		}
+
+		if !match {
```
Won't this always return false if there's no match found in first full iteration over `dbScopes`?

sakjur replied:
Yes, right now it's defined as `scope AND scope AND scope AND scope`, so that should be correct, right?

sakjur replied:
(fwiw, this entire file will be rewritten with a more reasonable API, this is just here to support early testing)

wbrowne reviewed (approved):
LGTM!

marefr commented on pkg/services/accesscontrol/manager/manager.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+// AccessControlService is the service implementing role based access control.
+type AccessControlService struct {
```
Nit. Package and filename is manager but here you have AccessControlService

marefr reviewed (approved):
LGTM looks great :+1: Added some comments/suggestions, but no blockers really.

sakjur commented on pkg/services/accesscontrol/accesscontrol.go:
```diff
@@ -0,0 +1,44 @@
+	Evaluate(ctx context.Context, user *models.SignedInUser, permission string, scope ...string) (bool, error)
+}
+
+type Store interface {
```
I agree on 2 and 3 (or partly on 3, I think something that is `GetX(ctx context.Context, orgID, ID int64)` is better left without that opening, but anything more complicated than that would benefit from a change). That being said: I'll merge without fixing this for now, this PR is big enough already and I need to focus on getting some new PRs up for this scope as well.

For 1, I think this is correctly named this interface maps out the low-level storage interface. The evaluator is the part of the service that most other services will interact with.

marefr replied:
> For 1, I think this is correctly named this interface maps out the low-level storage interface. The evaluator is the part of the service that most other services will interact with.

Oh I see. Didn't understand that :+1: So the idea would to have different implementations of the store to support "simple"/"advanced" permissions?

sakjur replied:
Yup. If you don't need the complexity, you can test towards a fake `Evaluator`, which will be a lot easier to implement.

sakjur committed:
Update pkg/services/accesscontrol/database/database_test.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

sakjur committed:
Update pkg/services/accesscontrol/middleware/middleware.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

sakjur commented on pkg/services/accesscontrol/manager/evaluator.go:
```diff
@@ -0,0 +1,72 @@
+		roles = append(roles, string(role))
+	}
+	if user.IsGrafanaAdmin {
+		roles = append(roles, "Grafana Admin")
```
:+1: That'd make sense

sakjur commented on pkg/services/accesscontrol/middleware/middleware.go:
```diff
@@ -0,0 +1,47 @@
+				scopes[i] = buf.String()
+			}
+
+			hasAccess, err := ac.Evaluate(context.TODO(), c.SignedInUser, permission, scopes...)
```
Thank you, I hadn't figured out how to get a proper context from Macaron's contexts.

sakjur commented on pkg/services/accesscontrol/models.go:
```diff
@@ -0,0 +1,188 @@
+
+// Role is the model for Role in RBAC.
+type Role struct {
+	ID          int64  `json:"id" xorm:"pk autoincr 'id'"`
+	OrgID       int64  `json:"orgId" xorm:"org_id"`
+	Version     int64  `json:"version"`
+	UID         string `xorm:"uid" json:"uid"`
```
Yes, because otherwise it'll serialize to `u_i_d`. Most other places still calls things like this `Uid` or `OrgId`.

sakjur committed:
Merge branch 'master' into access-control-merge-to-master

sakjur committed:
Fix tests, rename service->manager

sakjur committed:
Introduce constant for "Grafana Admin"
